resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java

```diff
@@ -406,30 +459,67 @@ private void onEvent(final InboundSseEvent event)
          {
             request.header(SseConstants.LAST_EVENT_ID_HEADER, lastEventId);
          }
-         if (disableKeepAlive)
-         {
-            request.header(HttpHeaders.CONNECTION, "close");
-         }
+         // We have to set Connection header to 'close' in order to
+         // disable persistent connection because of Apache HttpClient.
+         // Explanation:
+         // When dealing with persistent connections the connection
+         // provider implementation (Apache HttpClient or JAVA
+         // HttpUrlConnection) will try to reuse them when it is
+         // possible.
+         // One condition for a persistent connection to be reused is
+         // that it content should have been fully consumed
+         // (inputStream.read()=-1) else the connection won't be reused.
+         // Both Apache HttpClient and JAVA HttpUrlConnection advise us
+         // when it is possible to consume the whole response content
+         // by reading the inputStream and once done to invoke
+         // inputStream.close().
+         // However in many case, we are not interested in reading the
+         // whole response content and we just want to close the
+         // inputStream right away even if the remote side has not finish
+         // pushing content (SseEventSink not close on server side for
+         // example).
+         // - JAVA HttpUrlConnection allows us to invoke
+         // inputStream.close() to do so in a non blocking fashion (it
+         // will try to read the whole content in hurry mode without
+         // blocking. If it succeeds the connection will be reused else
+         // it will be discarded).
+         // - At the opposite invoking inputStream.close() on the Apache
+         // HTTPClient will block until the whole response content has
+         // been read. And depending on the remote side we can be blocked
+         // for really long...
+         // One solution provided by Apache HTTPClient to do so is
+         // invoking CloseableHttpResponse.close() instead. But it will
+         // not try to do best effort like JAVA HttpUrlConnection, it
+         // will discard the connection even if the whole response
+         // content can be read.
+         // http://hc.apache.org/httpcomponents-client-4.4.x/tutorial/html/fundamentals.html#d5e145
+         // https://docs.oracle.com/javase/8/docs/technotes/guides/net/http-keepalive.html
+         request.header(HttpHeaders.CONNECTION, "close");
          return request;
```

Comment:
@NicoNes  I still don't understand why it has to disable keep alive. How can we enable it ?

Reply:
Hi @jimma 

To be simple, without disabling keep alive, calling `response.close()` in `close()` method will block until the connection is closed on remote side.

That's what I'm testing in `SseEventSourceTest.testCloseSseEventSource()`.
If you take a look at the remote side, the connection is never closed in `SseSmokeResource.sentEventAndStayOpen()`. On the client side if keep alive was enabled `sseEventSource.close()` invoked by the try with resource will block indefinitely waiting for the remote side to close the connection.

The reason of such behavior can be found in `org.apache.http.impl.io.ChunkedInputStream` documentation:

>  Note that this class NEVER closes the underlying stream, even when close
 gets called.  Instead, it will read until the "end" of its chunking on
 close, which allows for the seamless execution of subsequent HTTP 1.1
 requests, while not requiring the client to remember to read the entire
 contents of the response.

And looking at the `org.apache.http.impl.io.ChunkedInputStream#close()` method we can see the blocking loop:

```
    @Override
    public void close() throws IOException {
        if (!closed) {
            try {
                if (!eof && state != CHUNK_INVALID) {
                    // read and discard the remainder of the message
                    final byte buff[] = new byte[BUFFER_SIZE];
        => blocking loop           while (read(buff) >= 0) {
                    }
                }
            } finally {
                eof = true;
                closed = true;
            }
        }
    }
```

So disabling keep alive allows to close response rigth away without blocking. Since the connection is not a persistent one and will not be reused, Apache HttpClient will not try to read the whole response content and will not block (that what I was trying to explain in the big comment sorry if it was not clear).